geopandas/sindex.py

```diff
@@ -70,6 +89,129 @@ def __init__(self, geometry):
+                often sorted but there is no guarantee.
+            Returns
+            -------
+            matches : ndarray of shape (n_results, )
+                Integer indices for matching geometries from the spatial index.
+            """  # noqa: E501
+            # handle invalid predicates
+            if predicate not in self.valid_query_predicates:
+                raise ValueError(
+                    "Got `predicate` = `%s`; " % predicate
+                    + "`predicate` must be one of %s" % self.valid_query_predicates
+                )
+            # handle empty / invalid geometries
+            if geometry is None:
+                # this is the behavior in pygeos.strtree.query, we mimic it here
+                return np.array([], dtype=np.intp)
+            if not isinstance(geometry, BaseGeometry):
+                raise TypeError(
+                    "Got `geometry` of type `%s`, " % type(geometry)
+                    + "`geometry` must be a shapely geometry."
+                )
+            if geometry.is_empty:
+                return np.array([], dtype=np.intp)
+
+            # get bounds
+            bounds = geometry.bounds  # rtree operates on bounds
+            tree_query = list(self.intersection(bounds, objects=False))
+
+            if not tree_query:
+                return np.array([], dtype=np.intp)
+
+            # check predicate
+            if predicate in ("intersects", "within"):
+                # only contains and intersects are supported by
+                # prepared geometries, see note below regarding within
+                res = []
+                if predicate == "within":
+                    # since these are inverse, we can flip the operation
+                    # and test with prepared predicates from tree
+                    predicate = "contains"
+                for i in tree_query:
+                    if self._prepared_geometries[i] is None:
+                        # if not already prepared, prepare and cache
+                        self._prepared_geometries[i] = prep(self._geometries[i])
+                    if getattr(self._prepared_geometries[i], predicate)(geometry):
+                        res.append(i)
+                tree_query = res
+            elif predicate == "contains" and len(tree_query) > 1:
+                # prepare this geometry
+                geometry = prep(geometry)
+                tree_query = [
+                    i
+                    for i in tree_query
+                    if getattr(geometry, predicate)(self._geometries[i])
+                ]
+            elif predicate is not None:
+                tree_query = [
+                    i
+                    for i in tree_query
+                    if getattr(geometry, predicate)(self._geometries[i])
+                ]
+
+            if not sort:
+                return np.array(tree_query)  # unsorted
```

Comment:
```suggestion
                return np.array(tree_query, dtype=np.intp)  # unsorted
```

? You do that for the empty arrays.

(and the same for the other occurrences that create an array)

Reply:
Good catch, will do.